### PR TITLE
feat: add results sorting by date (asc, desc) and task

### PR DESCRIPTION
**Pull request by RSMNYS**
https://github.com/mlcommons/mobile_app_open/issues/544.

Added sorting button (upward, downward arrows) for the benchmark results.
IMG_8305
IMG_8306

### CODE REVIEW

**Review by anhappdev — CHANGES_REQUESTED** [review]
1. Please use a descriptive title for the pull request. The title will be shown as a commit message, so it's recommended to use a format like this one: https://www.conventionalcommits.org
2. I think you can add other sorting options to this PR, so it's easier to test them all at once.

**Comment by RSMNYS** [comment]
I've changed this part a little bit by adding the sorting option to the filter screen itself. I think this is more user friendly. Still waiting for the info regards task sorting. 
IMG_8339
IMG_8340

**Comment by RSMNYS** [comment]
MicrosoftTeams-image (4)
MicrosoftTeams-image (3)


Guys, please suggest what other info we should display for the task throughput case? Maybe date? Or we can leave this for the UX guy.

File: flutter/lib/app_constants.dart [hunk 1]
```diff
@@ -108,3 +108,15 @@ class BackendId {
     apple,
   ];
 }
+
+class SortBy {
+  static const dateAsc = 'Date: from oldest to most recent';
```
Comment (anhappdev) [inline flutter/lib/app_constants.dart:113]: Please use `flutter/lib/l10n/app_en.arb` for UI text.

File: flutter/lib/ui/history/result_list_screen.dart [hunk 2]
```diff
@@ -54,23 +99,62 @@ class _ResultListScreenState extends State<ResultListScreen> {
...
+        .toList();
+  }
+
+  List<ListItem> _extendedResults(
+      List<ExtendedResult> items, ResultFilter filter) {
+    if (filter.sortBy == SortBy.dateAsc) {
+      items.sort((a, b) => a.meta.creationDate.compareTo(b.meta.creationDate));
+    } else if (filter.sortBy == SortBy.dateDesc) {
+      items.sort((a, b) => b.meta.creationDate.compareTo(a.meta.creationDate));
+    }
```
Comment (anhappdev) [inline flutter/lib/ui/history/result_list_screen.dart:148]: Can you extract all the sorting logics to a separated class, similar to how `class ResultFilter` works?
That makes it easier to extend the sorting options in the feature.

**Review by anhappdev — CHANGES_REQUESTED** [review]
> 1. What other info we should display for the task throughput case? Maybe date? Or we can leave this for the UX guy.

I would show: date, benchmark name, score as basis information.
And maybe backend name, delegate name, accelerator name if the UX guy can fit them nicely in the screen.

**Comment by freedomtan** [comment]
For what other to show. We need to bring this to the mobile working group meeting and ask David to comment here.

**Comment by anhappdev** [comment]
Currently, the iOS test failed because of an open issue related to Homebrew (see https://github.com/Homebrew/homebrew-core/issues/135315)

Please cherry-pick this commit https://github.com/mlcommons/mobile_app_open/pull/746/commits/80b99c59bee9afd793c0c5be4ce912c26a8de37f to fix it in your PR.

File: flutter/lib/ui/history/result_filter_screen.dart [hunk 3]
```diff
@@ -93,30 +97,43 @@ class _ResultFilterScreenState extends State<ResultFilterScreen> {
...
-        onChanged: (value) => setState(() {
-              filter.platform = value!;
+        choices: EnvPlatform.values
+            .map((e) => DropdownOption(e.name, e.name))
+            .toList(),
+        value: filter.platform != null
+            ? DropdownOption(filter.platform, filter.platform ?? '')
+            : null,
+        onChanged: (option) => setState(() {
+              filter.platform = option?.value as String;
```
Comment (anhappdev) [inline flutter/lib/ui/history/result_filter_screen.dart:107]: The app crashed when setting platform filter and sorting by task throughput.
```
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following _CastError was thrown building ResultListScreen(dirty, dependencies:
[_InheritedProviderScope<BenchmarkState?>, _LocalizationsScope-[GlobalKey#ae549]], state:
_ResultListScreenState#353e7):
Null check operator used on a null value

The relevant error-causing widget was:
  ResultListScreen
  ResultListScreen:file:///Users/anh/dev/mlcommons/mobile_app_open/flutter/lib/ui/run/app_bar.dart:20:59

When the exception was thrown, this was the stack:
#0      ResultSort.sortBenchmarks.<anonymous closure> (package:mlperfbench_common/data/result_sort.dart:32:46)
#1      Sort._insertionSort (dart:_internal/sort.dart:69:36)
#2      Sort._doSort (dart:_internal/sort.dart:58:7)
#3      Sort.sort (dart:_internal/sort.dart:33:5)
#4      ListMixin.sort (dart:collection/list.dart:356:10)
#5      ResultSort.sortBenchmarks (package:mlperfbench_common/data/result_sort.dart:32:15)
#6      ResultsDataProvider.resultItems (package:mlperfbench/ui/history/results_data_provider.dart:26:25)
#7      _ResultListScreenState.build (package:mlperfbench/ui/history/result_list_screen.dart:35:56)
#8      StatefulElement.build (package:flutter/src/widgets/framework.dart:5080:27)
#9      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4968:15)
#10     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5133:11)
#11     Element.rebuild (package:flutter/src/widgets/framework.dart:4690:5)
#12     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2743:19)
#13     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:863:21)
#14     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:381:5)
#15     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1289:15)
#16     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1218:9)
#17     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1076:5)
#18     _invoke (dart:ui/hooks.dart:145:13)
#19     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:338:5)
#20     _drawFrame (dart:ui/hooks.dart:112:31)

════════════════════════════════════════════════════════════════════════════════════════════════════
```

File: flutter/lib/ui/history/result_filter_screen.dart [hunk 4]
```diff
@@ -174,6 +191,23 @@ class _ResultFilterScreenState extends State<ResultFilterScreen> {
     );
   }
 
+  Widget _sortBy(ResultFilter filter) {
```
Comment (anhappdev) [inline flutter/lib/ui/history/result_filter_screen.dart:194]: The sort should not be dependent on the filter or part of the filter.

File: flutter/lib/ui/history/result_list_item.dart [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import 'package:flutter/material.dart';
+
+import 'package:mlperfbench_common/data/extended_result.dart';
+import 'package:mlperfbench_common/data/results/benchmark_result.dart';
+
+import 'package:mlperfbench/localizations/app_localizations.dart';
+import 'package:mlperfbench/ui/history/utils.dart';
+import 'list_item.dart';
```
Comment (anhappdev) [inline flutter/lib/ui/history/result_list_item.dart:8]: Please use the same import style (package in this case). Please also double-check other files in this PR for the same import style.

File: flutter/lib/ui/history/result_list_screen.dart [hunk 6]
```diff
@@ -50,27 +56,44 @@ class _ResultListScreenState extends State<ResultListScreen> {
...
   }
+
+  List<ListItem> _listItems(List<ResultItem> resultItems) {
+    return resultItems.map((resultItem) {
+      if (resultItem.item is BenchmarkExportResult) {
+        return _benchmarkListItem(resultItem.item as BenchmarkExportResult);
+      }
+      return _resultListItem(resultItem.item as ExtendedResult);
+    }).toList();
+  }
```
Comment (anhappdev) [inline flutter/lib/ui/history/result_list_screen.dart:76]: The UI for the list of past results should remain consistent in terms of its level of detail, regardless of whether a sort option is applied.

File: flutter_common/lib/data/result_filter.dart [hunk 7]
```diff
@@ -16,6 +17,7 @@ class ResultFilter {
   String? manufacturer;
   String? soc;
   String? benchmarkId;
+  SortByValues? sortBy;
```
Comment (anhappdev) [inline flutter_common/lib/data/result_filter.dart:20]: The sort should not be dependent on the filter or part of it.

File: flutter_common/lib/data/result_sort.dart [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  static List<BenchmarkExportResult> sortBenchmarks(
+      SortByValues sortBy, List<BenchmarkExportResult> items) {
+    List<BenchmarkExportResult> sortedItems =
+        List<BenchmarkExportResult>.from(items);
+    switch (sortBy) {
+      case SortByValues.taskThroughputDesc:
+        items.sort((a, b) => b.performanceRun!.throughput!
+            .compareTo(a.performanceRun!.throughput!));
```
Comment (anhappdev) [inline flutter_common/lib/data/result_sort.dart:33]: Please avoid using force unwrapping.

**Review by anhappdev — CHANGES_REQUESTED** [review]
Please add a unit test for the sorting feature.

File: flutter/lib/ui/history/result_filter_screen.dart (hunk 4 above)
Comment (RSMNYS) [reply flutter/lib/ui/history/result_filter_screen.dart:194]: Should we move then the sort by option out from the filter screen? And introduce the new sort screen with its state?

Comment (RSMNYS) [reply flutter/lib/ui/history/result_filter_screen.dart:194]: Generally I think in our case the filtering and sorting are related and used together. We don't use the sorting separately in the other part of the app. Ofcourse we can separate it (it's easy to do, after the refactoring), but don't think it makes a lot of sense currently. If in the future we have the need to use it separately or have the more advanced sorting we can separate. Please let me know what do you think.

Comment (anhappdev) [reply flutter/lib/ui/history/result_filter_screen.dart:194]: > Specifically: Show the list of benchmark-specific results in the history screen (one level deeper than currently). Is this still valid?

Yes. It's still valid. The history screen should show a list of benchmark-specific results.

> now you are saying to not change the list's structure.

Not sure where I mentioned that. Maybe a misunderstanding?

Comment (RSMNYS) [reply flutter/lib/ui/history/result_filter_screen.dart:194]: Ok, now is clear. So for all cases we need to show results by benchmarks, not only for the sorting by task throughput case. Correct? Was thinking we need to show 2 distinguished lists for each case separately. Thanks for the clarification. Will apply changes.

File: flutter/lib/ui/history/result_filter_screen.dart (hunk 3 above)
Comment (RSMNYS) [reply flutter/lib/ui/history/result_filter_screen.dart:109]: Can't reproduce anymore

File: flutter/lib/ui/history/result_list_screen.dart (hunk 2 above)
Comment (RSMNYS) [reply flutter/lib/ui/history/result_list_screen.dart]: Refactored

File: flutter/lib/app_constants.dart [hunk 9]
```diff
@@ -109,6 +113,24 @@ class BackendId {
   ];
 }
 
+class SortByItem {
+  String label;
+  SortByValues value;
+
+  SortByItem(this.label, this.value);
```
[the remark this answers is not in the document]
Comment (RSMNYS) [reply flutter/lib/app_constants.dart]: Moved as a separate class to the flutter_common.

File: flutter/lib/ui/history/result_filter_screen.dart (hunk 4 above)
Comment (RSMNYS) [reply flutter/lib/ui/history/result_filter_screen.dart]: Moved sort logic to a separate class;

File: flutter/lib/ui/history/result_list_item.dart (hunk 5 above)
Comment (RSMNYS) [reply flutter/lib/ui/history/result_list_item.dart:9]: docker/lint fixed that

File: flutter/lib/ui/history/result_list_screen.dart (hunk 6 above)
Comment (RSMNYS) [reply flutter/lib/ui/history/result_list_screen.dart]: Now in the result list we show only BenchmarkExportResult

File: flutter_common/lib/data/result_filter.dart (hunk 7 above)
Comment (RSMNYS) [reply flutter_common/lib/data/result_filter.dart]: Refactored

File: flutter/lib/ui/history/result_list_item.dart (hunk 5 above)
Comment (anhappdev) [reply flutter/lib/ui/history/result_list_item.dart:9]: What I mean is that `import 'list_item.dart';` should be`import 'package:mlperfbench/ui/history/list_item.dart';`. But it's ok for now. I will change it and add a lint rule for this later.

**Review by anhappdev — APPROVED** [review]
The code works. The UI needs to be improved but let do it in another PR.

**Pull request merged**